modules/nf-cid/src/main/nextflow/data/cid/CidObserver.groovy

```diff
@@ -0,0 +1,415 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package nextflow.data.cid
+
+import static nextflow.data.cid.fs.CidPath.*
+
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.attribute.BasicFileAttributes
+import java.time.Instant
+
+import groovy.transform.CompileStatic
+import groovy.util.logging.Slf4j
+import nextflow.Session
+import nextflow.data.cid.model.Checksum
+import nextflow.data.cid.model.DataOutput
+import nextflow.data.cid.model.DataPath
+import nextflow.data.cid.model.Parameter
+import nextflow.data.cid.model.TaskOutputs
+import nextflow.data.cid.model.Workflow
+import nextflow.data.cid.model.WorkflowOutputs
+import nextflow.data.cid.model.WorkflowRun
+import nextflow.file.FileHelper
+import nextflow.file.FileHolder
+import nextflow.processor.TaskHandler
+import nextflow.processor.TaskRun
+import nextflow.script.ScriptMeta
+import nextflow.script.params.DefaultInParam
+import nextflow.script.params.FileInParam
+import nextflow.script.params.FileOutParam
+import nextflow.script.params.InParam
+import nextflow.script.params.OutParam
+import nextflow.trace.TraceObserver
+import nextflow.trace.TraceRecord
+import nextflow.util.CacheHelper
+import nextflow.util.PathNormalizer
+import nextflow.util.TestOnly
+/**
+ * Observer to write the generated workflow metadata in a CID store.
+ *
+ * @author Paolo Di Tommaso <paolo.ditommaso@gmail.com>
+ */
+@Slf4j
+@CompileStatic
+class CidObserver implements TraceObserver {
+
+    private String executionHash
+    private CidStore store
+    private Session session
+    private WorkflowOutputs workflowResults
+    private Map<String,String> outputsStoreDirCid = new HashMap<String,String>(10)
+
+    CidObserver(Session session, CidStore store){
+        this.session = session
+        this.store = store
+    }
+
+    @Override
+    void onFlowCreate(Session session) {
+        this.store.getHistoryLog().write(session.runName, session.uniqueId, '-')
+    }
+
+    @TestOnly
+    String getExecutionHash(){ executionHash }
+
+    @Override
+    void onFlowBegin() {
+        executionHash = storeWorkflowRun()
+        final executionUri = asUriString(executionHash)
+        workflowResults = new WorkflowOutputs(
+            Instant.now(),
+            executionUri,
+            new HashMap<String, Object>()
+        )
+        this.store.getHistoryLog().updateRunCid(session.uniqueId, executionUri)
+    }
+
+    @Override
+    void onFlowComplete(){
+        if (this.workflowResults){
+            workflowResults.createdAt = Instant.now()
+            final key = executionHash + SEPARATOR + 'outputs'
+            this.store.save(key, workflowResults)
+        }
+    }
+
+    protected String storeWorkflowRun() {
+        final normalizer = new PathNormalizer(session.workflowMetadata)
```

Comment:
Likely this can be a class attribute to avoid creating a new instance for each task, see below